### ISSUE TITLE
Over-estimate the loss from ALL action space?

### ISSUE THREAD

**Issue by QuantHao** [PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook#32]
Hi, @praveen-palanisamy 

Using tf.keras.Model.fit and tf.keras.Model.predict instead of low level API such as tf.GradientTape in Chapter 3 makes your code very concise and easy to read.

However, I am confused about TD error which I GUESS it is over-estimated for the following reason:

In Chapter 2, you only compute TD error for state-action pairs (s, a) that **actually appear** (by that I mean q[state][action])
https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/997f7e3bf1ca65baa21c9d44e5c5d64c4f685b20/Chapter02/6_q_learning.py#L28-L31

In Chapter 3, you compute TD error for **ALL** state-action pairs, regardless of appears or not, since _targets_ is a tensor with shape (batch_size, action_dim) instead of (batch_size, 1)
https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/997f7e3bf1ca65baa21c9d44e5c5d64c4f685b20/Chapter03/1_dqn.py#L110-L115

Therefore, I think total TD error is over-estimated. My way to fix it is to replace initialize _targets_ 
https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/997f7e3bf1ca65baa21c9d44e5c5d64c4f685b20/Chapter03/1_dqn.py#L110
with _targets = self.model.predict(states)_, then update with _next_q_values_ as your current code.

In this way, targets is NON-ZERO only for state-action pairs **actually appear**. I hope my description is not too lengthy and makes sense to you.

Looking forward to your reply. Thanks in advance.

**Comment by praveen-palanisamy** [comment]
Hi @QuantHao ,
I may be misunderstanding you but please consider the following to see if your concern/confusion is addressed:

On line 109 in the following snippet from the DQN recipe you linked in your question:
https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/997f7e3bf1ca65baa21c9d44e5c5d64c4f685b20/Chapter03/1_dqn.py#L109-L115

The experience (transitions) sampled from the replay buffer _only_ contains the actions taken by the DQN agent. We are then using the target Q network's parameters to get the target Q values for those actions in the next few lines in order to train the main DQN. Does this make sense?

I think you got confused with the action dimension (which is 2 for the default `CartPole-v0` env used in [Chapter03/1_dqn.py](https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/997f7e3bf1ca65baa21c9d44e5c5d64c4f685b20/Chapter03/1_dqn.py)) and the possible values per action as in the case of tabular Q-learning.

Did the above address your concern/confusion or did I miss something?

**Comment by QuantHao** [comment]
Hi, praveen

Thanks a lot for your reply. But I still cannot convince myself.

> The experience (transitions) sampled from the replay buffer only contains the actions taken by the DQN agent.

(1) Yes, I agree. Only action (0 or 1 out of Discrete(2)) is taken and stored https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/997f7e3bf1ca65baa21c9d44e5c5d64c4f685b20/Chapter03/1_dqn.py#L123-L127

> We are then using the target Q network's parameters to get the target Q values for **those actions**

(2) Well, I think `targets` contains ALL possible actions rather than actions taken by DQN agent because https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/997f7e3bf1ca65baa21c9d44e5c5d64c4f685b20/Chapter03/1_dqn.py#L110
`targets` is a tensor with shape (batch size, 2), given action_dim is 2.

(3) But I do agree you only update `targets` where actions **actually taken** by agent https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/997f7e3bf1ca65baa21c9d44e5c5d64c4f685b20/Chapter03/1_dqn.py#L111-L114

(Summary) So, here is the core of porblem: since `targets` is computed by `self.target_model.predict(states)`, whose parameters are different from `self.model`. Therefore, locations of  `targets` where action NOT taken by agent also have NON-ZERO error during training https://github.com/PacktPublishing/Tensorflow-2-Reinforcement-Learning-Cookbook/blob/997f7e3bf1ca65baa21c9d44e5c5d64c4f685b20/Chapter03/1_dqn.py#L115

If `targets` is initialize by `self.model.predict(states)`  instead, and update as (3). Then locations of  `targets` where action NOT taken have zero TD error.

I hope above explain myself a little better.

**Comment by praveen-palanisamy** [comment]
Hey @QuantHao ,
I appreciate your detailed clarification! I see your point clearly. 
You are right. The initialization of `targets` should be with `self.model.predict(states)` to make sure there's non-zero TD-error/gradients flowing through the q-value _only_ for the action chosen by the current/behavior policy.

I'll fix this bug in an update soon.
